### PR TITLE
Added some more reference tests for Buffer<T> and ROBuffer<T>

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
@ahsonkhan, @shiftylogic, @joshfree

### CODE REVIEW

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 1]
```diff
@@ -237,5 +243,37 @@ static void BufferLifetime(Buffer<byte> buffer)
             //{
             //}
         }
+
+        static void BufferBasics(ReadOnlyBuffer<byte> buffer)
+        {
+            var span = buffer.Span;
+            Assert.Equal(buffer.Length, span.Length);
+            Assert.True(buffer.IsEmpty || buffer.Length != 0);
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: Why do we need this check?

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 2]
```diff
@@ -237,5 +243,37 @@ static void BufferLifetime(Buffer<byte> buffer)
...
+
+            var slice = buffer.Slice(1);
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: if buffer.IsEmpty is true, won't this result in a buffer of length -1?

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 3]
```diff
@@ -237,5 +243,37 @@ static void BufferLifetime(Buffer<byte> buffer)
...
+        {
+            var array = buffer.ToArray();
+            using(var handle = buffer.Retain())
+            using (var pinned = buffer.Pin())
+            {
+                unsafe
+                {
+                    var p = (byte*)pinned.PinnedPointer;
+                    Assert.True(null != p);
+                    for(int i=0; i<buffer.Length; i++)
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: nit: spacing.

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 4]
```diff
@@ -237,5 +243,37 @@ static void BufferLifetime(Buffer<byte> buffer)
...
+            for (int i = 0; i < slice.Length; i++) Assert.Equal(slice.Span[i], span[i+1]);
+
+            var array = buffer.ToArray();
+            for (int i = 0; i < array.Length; i++) Assert.Equal(array[i], span[i]);
+        }
+
+        static void BufferLifetime(ReadOnlyBuffer<byte> buffer)
+        {
+            var array = buffer.ToArray();
+            using(var handle = buffer.Retain())
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: nit: spacing.

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: But the test can run on zero length buffers. This test ensures that is buffer is not empty, then Length !=0

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs (hunk 2 above)
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: good point. I wonder why the tests don't fail. I will investigate.

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs (hunk 3 above)
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: will fix

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs (hunk 4 above)
Comment (KrzysztofCwalina) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: will fix

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs (hunk 1 above)
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: Why are we checking both buffer.IsEmpty and buffer.Length?
IsEmpty already does the check if Length == 0.
```C#
public bool IsEmpty => Length == 0;
```

https://github.com/dotnet/corefxlab/blob/master/src/System.Buffers.Primitives/System/Buffers/Buffer.cs#L61

> This test ensures that is buffer is not empty, then Length !=0

How does this test ensure that?

Assert.True(buffer.IsEmpty || X) => passes regardless of what X returns. If we have buffer.Length != 0, buffer.Empty should be false, No? Am I missing something here?

File: tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs [hunk 5]
```diff
@@ -202,35 +210,39 @@ static void BufferBasics(Buffer<byte> buffer)
             var span = buffer.Span;
             Assert.Equal(buffer.Length, span.Length);
             Assert.True(buffer.IsEmpty || buffer.Length != 0);
+            Assert.True(!buffer.IsEmpty || buffer.Length == 0);
```
Comment (ahsonkhan) [inline tests/System.Buffers.Primitives.Tests/BufferReferenceTesting.cs]: Is the following the intention of the two Asserts?
```C#
if (buffer.Empty)
{
    Assert.Equal(0, buffer.Length);
}
else
{
    Assert.True(buffer.Length != 0)
}
```

**Pull request merged**